Add tab blocks to the blocks product editor

joshuatf opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This adds a `tab` block for use in the new product editor that is treated like a normal block.  The tab button itself is slotfilled above the editor, while the selected tab is set using context.

This does not address the following, which will be added in follow-up PRs:

* Accessibility
* Unit tests
* Placement within the header (new header has not yet been merged)
<img width="769" alt="Screen Shot 2023-03-10 at 8 55 54 AM" src="https://user-images.githubusercontent.com/10561050/224376831-e7fc73d4-0b9e-49dd-95e9-464d5192fed8.png">



Closes (part of) #37096  .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Please include detailed instructions on how these changes can be tested, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Check out `add/37096-test` for initially hydrated sample tab blocks
2. Make sure that the blocks product editor feature is enabled
3. Navigate to the add product page
4. See that the "General" tab is initially selected and the general content is shown
5. Click "Shipping" and see that the shipping tab is selected and shipping content is shown
6. Refresh the page (noting `&tab=shipping` is in the URL) and note that shipping is initially selected
7. Note that styling is a bit off since the new header is not yet in place, but margin/padding matches the Figma designs

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Add initial tab block and tabs render

joshuatf committed:
Add tabs styling

joshuatf committed:
Set initially selected tab on render

joshuatf committed:
Set initial tab based on query param

joelclimbsthings commented on packages/js/product-editor/src/components/tab/block.json:
```diff
@@ -0,0 +1,27 @@
+	"supports": {
+		"align": false,
+		"html": false,
+		"multiple": false,
```
I believe having this set to `false` is giving me this error:

![image](https://user-images.githubusercontent.com/444632/224424268-2354cdd6-3e74-432c-ac50-a44c284b2bb9.png)

I think this should be `true` ?

That said, this would be true of some of our other blocks such as `Section`, but that doesn't seem to be giving me the same error.

joshuatf replied:
Good catch.  I never ran into this with this block already inserted, but regardless seems like something that should be enabled. Updated in https://github.com/woocommerce/woocommerce/pull/37174/commits/ebfafabc25a1e0a770c6eec885e4364278cbaaa0

joelclimbsthings commented on packages/js/product-editor/src/components/tabs/tabs.tsx:
```diff
@@ -0,0 +1,84 @@
+
+	function onClick( tabId: string ) {
+		window.document.documentElement.scrollTop = 0;
+		navigateTo( {
```
Nice use of this util 👍🏻

joelclimbsthings commented on packages/js/product-editor/src/components/tab/block.json:
```diff
@@ -0,0 +1,27 @@
+	"keywords": [ "products", "tab", "group" ],
+	"textdomain": "default",
+	"attributes": {
+		"id": {
```
Would it be useful to have support for `disabled` via an attribute here?

joshuatf replied:
Definitely.  Planned for the follow-ups to tabs, but I'd like for us to flesh out how we plan on "templating" different product types.

If we go with individual templates per product type, attributes may work well, but otherwise we may have to build logic directly into the block.

joelclimbsthings reviewed (changes requested):
Thanks much @joshuatf , and well done. Had a comment or two, and also have this in the console that I'm wondering if we should address:

![image](https://user-images.githubusercontent.com/444632/224432948-2f526057-22c2-485c-822e-a395e0444c2b.png)

joshuatf committed:
Add template placeholder with tabs

joshuatf committed:
Add changelog entry

joshuatf committed:
Allow multiple tab and section blocks

joshuatf committed:
Add changelog entry for wc

joshuatf committed:
Fix up arrow alignment

joshuatf commented:
Thanks for the feedback, @joelclimbsthings!  Addressed the comments.

>Had a comment or two, and also have this in the console that I'm wondering if we should address:

I can create a follow-up for this.  I don't see this being registered in WC blocks either, but they use the same category 🤔 

We might need to double check before adding it to avoid duplicate categories.

joshuatf committed:
Fix up block configuration type

joelclimbsthings reviewed (dismissed):
Thanks @joshuatf , changes look good and everything is working well. 🚢

joshuatf committed:
Add missing navigation dependency to product editor
